File: ios/Sources/AuthenticationPlugin/AuthenticationPlugin.swift
import Capacitor
import CommonCrypto
import Foundation
import TruvideoSdk

@objc(AuthenticationPlugin)
public class AuthenticationPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AuthenticationPlugin"
    public let jsName = "Authentication"
    
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "echo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(
            name: "isAuthenticated", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(
            name: "isAuthenticationExpired", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(
            name: "generatePayload", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(
            name: "authenticate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(
            name: "authenticateWithOtp", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(
            name: "generateOtp", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(
            name: "initAuthentication", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(
            name: "clearAuthentication", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(
            name: "toSha256String", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sha256", returnType: CAPPluginReturnPromise),
    ]
    private let implementation = Authentication()
    private var isConfigured = false
    
    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        print("[AuthenticationPlugin] Echo called with value: \(value)")
        call.resolve(["value": value])
    }
    
    private func configureSDK() {
        if isConfigured {
            return
        }
        
        print("[AuthenticationPlugin] Auto-configuring SDK...")
        
        // Create TruVideoOptions object with default settings
        let truVideoOptions = TruVideoOptions()
        
        // Configure the SDK with options
        TruvideoSdk.configure(with: truVideoOptions)
        isConfigured = true
        print("[AuthenticationPlugin] SDK configured successfully")
    }
    
    private func ensureConfigured() -> Bool {
        if !isConfigured {
            // Automatically configure on first use
            configureSDK()
        }
        return true
    }
    
    @objc func isAuthenticated(_ call: CAPPluginCall) {
        print("[AuthenticationPlugin] isAuthenticated called")
        ensureConfigured()
        
        let isAuth = TruvideoSdk.isAuthenticated
        print("[AuthenticationPlugin] isAuthenticated result: \(isAuth)")
        call.resolve(["isAuthenticated": isAuth])
    }
    
    @objc func environment(_ call: CAPPluginCall) {
        //let isAuth = (try? TruvideoSdk.environment) ?? false
        print("[AuthenticationPlugin] isAuthenticated called. Result: \("")")
        call.resolve(["environment": ""])
    }
    
    @objc func version(_ call: CAPPluginCall) {
        //let isAuth = (try? TruvideoSdk.version) ?? false
        print("[AuthenticationPlugin] isAuthenticated called. Result: \("")")
        call.resolve(["version": ""])
    }
    
    @objc func getApiKey(_ call: CAPPluginCall) {
        //let isAuth = (try? TruvideoSdk.getApiKey()) ?? false
        print("[AuthenticationPlugin] isAuthenticated called. Result: ")
        call.resolve(["apiKey": ""])
    }
    
    
    @objc func isAuthenticationExpired(_ call: CAPPluginCall) {
        print("[AuthenticationPlugin] isAuthenticationExpired called")
        ensureConfigured()
        
        do {
            let isExpired = try TruvideoSdk.isAuthenticationExpired()
            print("[AuthenticationPlugin] isAuthenticationExpired result: \(isExpired)")
            call.resolve(["isAuthenticationExpired": isExpired])
        } catch {
            let errorMessage = "Failed to check if authentication is expired: \(error.localizedDescription)"
            print("[AuthenticationPlugin] isAuthenticationExpired error: \(errorMessage)")
            call.reject(errorMessage, nil, error)
        }
    }
    
    @objc func generatePayload(_ call: CAPPluginCall) {
        print("[AuthenticationPlugin] generatePayload called")
        ensureConfigured()
        
        do {
            let payload = try TruvideoSdk.generatePayload()
            print("[AuthenticationPlugin] generatePayload result: \(payload)")
            call.resolve(["generatePayload": payload])
        } catch {
            let errorMessage = "Failed to generate payload: \(error.localizedDescription)"
            print("[AuthenticationPlugin] generatePayload error: \(errorMessage)")
            call.reject(errorMessage, nil, error)
        }
    }
    
    
    
    @objc func authenticate(_ call: CAPPluginCall) {
        ensureConfigured()
        
        // Extract parameters - documentation mentions apiKey, secretKey, externalId
        // Current implementation uses apiKey, payload, signature, externalId
        // Adjust based on actual SDK signature
        guard let apiKey = call.getString("apiKey"),
              let payload = call.getString("payload"),
              let signature = call.getString("signature"),
              let externalId = call.getString("externalId") else {
            print("[AuthenticationPlugin] authenticate failed: Missing parameters")
            call.reject("Missing required parameters: apiKey, payload, signature, and externalId are required", "MISSING_PARAMETERS")
            return
        }
        
        print("[AuthenticationPlugin] authenticate called with apiKey: \(apiKey)")
        
        Task {
            do {
                try await TruvideoSdk.authenticate(apiKey: apiKey, payload: payload, signature: signature, externalId: externalId)
                print("[AuthenticationPlugin] authenticate success")
                call.resolve(["authenticate": "Authentication success"])
            } catch {
                // Handle authentication errors
                let errorMessage = "Authentication failed: \(error.localizedDescription)"
                let errorCode: String
                
                // Determine error code based on error description or type
                if error.localizedDescription.lowercased().contains("configuration") || 
                   error.localizedDescription.lowercased().contains("not configured") {
                    errorCode = "CONFIGURATION_REQUIRED"
                } else if error.localizedDescription.lowercased().contains("credential") ||
                          error.localizedDescription.lowercased().contains("invalid") ||
                          error.localizedDescription.lowercased().contains("failed") {
                    errorCode = "AUTHENTICATION_FAILED"
                } else {
                    errorCode = "AUTHENTICATION_ERROR"
                }
                
                print("[AuthenticationPlugin] authenticate error: \(errorMessage) (code: \(errorCode))")
                call.reject(errorMessage, errorCode, error)
            }
        }
    }

    /// Pure HTTP OTP generation (same contract as Android / React Native `generateOtp`).
    @objc func generateOtp(_ call: CAPPluginCall) {
        guard let baseUrl = call.getString("baseUrl"),
              let apiKey = call.getString("apiKey"),
              let secret = call.getString("secret"),
              let externalId = call.getString("externalId") else {
            call.reject("Missing baseUrl, apiKey, secret, or externalId", "OTP_GENERATE_ERROR", nil)
            return
        }

        let trimmedApiKey = apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        let trimmedSecret = secret.trimmingCharacters(in: .whitespacesAndNewlines)
        let trimmedExternalId = externalId.trimmingCharacters(in: .whitespacesAndNewlines)
        let trimmedBase = baseUrl.trimmingCharacters(in: .whitespacesAndNewlines)

        if trimmedApiKey.isEmpty {
            call.reject("apiKey cannot be empty", "OTP_GENERATE_ERROR", nil)
            return
        }
        if trimmedSecret.isEmpty {
            call.reject("secret cannot be empty", "OTP_GENERATE_ERROR", nil)
            return
        }
        if trimmedExternalId.isEmpty {
            call.reject("externalId cannot be empty", "OTP_GENERATE_ERROR", nil)
            return
        }
        if trimmedBase.isEmpty {
            call.reject("baseUrl cannot be empty", "OTP_GENERATE_ERROR", nil)
            return
        }

        print("[AuthenticationPlugin] generateOtp started")

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let cleanBase = Self.trimTrailingSlashes(trimmedBase)
                let endpoint = "\(cleanBase)/api/v1/auth/otp/generate"
                guard let url = URL(string: endpoint) else {
                    throw NSError(domain: "OTP_GENERATE_ERROR", code: 0, userInfo: [NSLocalizedDescriptionKey: "Invalid base URL"])
                }

                let bodyDict: [String: Any] = ["externalId": trimmedExternalId]
                let bodyData = try JSONSerialization.data(withJSONObject: bodyDict, options: [])
                guard let bodyString = String(data: bodyData, encoding: .utf8) else {
                    throw NSError(domain: "OTP_GENERATE_ERROR", code: 0, userInfo: [NSLocalizedDescriptionKey: "Failed to encode request body"])
                }

                guard let signature = Self.hmacSha256Hex(secret: trimmedSecret, payload: bodyString) else {
                    throw NSError(domain: "OTP_GENERATE_ERROR", code: 0, userInfo: [NSLocalizedDescriptionKey: "Failed to generate request signature"])
                }

                var request = URLRequest(url: url)
                request.httpMethod = "POST"
                request.timeoutInterval = 15
                request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                request.setValue(trimmedApiKey, forHTTPHeaderField: "x-authentication-api-key")
                request.setValue(signature, forHTTPHeaderField: "x-authentication-signature")
                request.httpBody = bodyData

                let semaphore = DispatchSemaphore(value: 0)
                var responseData: Data?
                var responseError: Error?
                var httpResponse: HTTPURLResponse?

                let task = URLSession.shared.dataTask(with: request) { data, response, error in
                    responseData = data
                    responseError = error
                    httpResponse = response as? HTTPURLResponse
                    semaphore.signal()
                }
                task.resume()
                semaphore.wait()

                if let err = responseError {
                    throw err
                }

                let status = httpResponse?.statusCode ?? -1
                let responseText = String(data: responseData ?? Data(), encoding: .utf8) ?? ""

                guard status >= 200, status <= 299 else {
                    let apiMsg = Self.parseJsonErrorMessage(responseText)
                    let message: String
                    if !apiMsg.isEmpty {
                        message = "OTP generate failed (\(status)): \(apiMsg)"
                    } else {
                        message = "OTP generate failed with status \(status)"
                    }
                    throw NSError(domain: "OTP_GENERATE_ERROR", code: status, userInfo: [NSLocalizedDescriptionKey: message])
                }

                var otpValue = ""
                if let obj = try? JSONSerialization.jsonObject(with: Data(responseText.utf8)) as? [String: Any] {
                    otpValue = (obj["otp"] as? String) ?? ""
                }
                if otpValue.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    throw NSError(domain: "OTP_GENERATE_ERROR", code: 0, userInfo: [NSLocalizedDescriptionKey: "OTP not found in response"])
                }

                print("[AuthenticationPlugin] generateOtp success")
                DispatchQueue.main.async {
                    call.resolve(["generateOtp": otpValue])
                }
            } catch {
                print("[AuthenticationPlugin] generateOtp error: \(error.localizedDescription)")
                DispatchQueue.main.async {
                    call.reject(error.localizedDescription, "OTP_GENERATE_ERROR", error)
                }
            }
        }
    }

    private static func trimTrailingSlashes(_ s: String) -> String {
        var r = s
        while r.hasSuffix("/") {
            r.removeLast()
        }
        return r
    }

    private static func hmacSha256Hex(secret: String, payload: String) -> String? {
        guard let keyData = secret.data(using: .utf8),
              let payloadData = payload.data(using: .utf8) else {
            return nil
        }
        var hmac = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        keyData.withUnsafeBytes { kb in
            payloadData.withUnsafeBytes { pb in
                CCHmac(CCHmacAlgorithm(kCCHmacAlgSHA256), kb.baseAddress, keyData.count, pb.baseAddress, payloadData.count, &hmac)
            }
        }
        return hmac.map { String(format: "%02x", $0) }.joined()
    }

    private static func parseJsonErrorMessage(_ responseText: String) -> String {
        guard let data = responseText.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return ""
        }
        let message = (obj["message"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if !message.isEmpty {
            return message
        }
        let detail = (obj["detail"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return detail
    }

    @objc func authenticateWithOtp(_ call: CAPPluginCall) {
        ensureConfigured()

        guard let otp = call.getString("otp") else {
            print("[AuthenticationPlugin] authenticateWithOtp failed: Missing otp parameter")
            call.reject("OTP is required", "AUTHENTICATION_FAILED", nil)
            return
        }

        let trimmedOtp = otp.trimmingCharacters(in: .whitespacesAndNewlines)

        guard !trimmedOtp.isEmpty else {
            print("[AuthenticationPlugin] authenticateWithOtp rejected: empty OTP")
            call.reject("OTP cannot be empty", "AUTHENTICATION_FAILED", nil)
            return
        }

        print("[AuthenticationPlugin] authenticateWithOtp called")

        Task {
            do {
                try await TruvideoSdk.authenticate(otp: trimmedOtp)

                let authenticated = TruvideoSdk.isAuthenticated
                guard authenticated else {
                    print("[AuthenticationPlugin] authenticateWithOtp: SDK reported not authenticated after OTP")
                    call.reject("OTP authentication failed", "AUTHENTICATION_OTP_FAILED", nil)
                    return
                }

                print("[AuthenticationPlugin] authenticateWithOtp success")
                call.resolve(["authenticateWithOtp": "Authentication successful"])
            } catch {
                let errorMessage = "OTP authentication failed: \(error.localizedDescription)"
                print("[AuthenticationPlugin] authenticateWithOtp error: \(errorMessage)")
                call.reject(errorMessage, "AUTHENTICATION_OTP_FAILED", error)
            }
        }
    }

    @objc func initAuthentication(_ call: CAPPluginCall) {
        print("[AuthenticationPlugin] initAuthentication called")
        ensureConfigured()
        
        Task {
            do {
                try await TruvideoSdk.initAuthentication()
                print("[AuthenticationPlugin] initAuthentication success")
                call.resolve(["initAuthentication": "Init success"])
            } catch {
                print("[AuthenticationPlugin] initAuthentication failed: \(error.localizedDescription)")
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func clearAuthentication(_ call: CAPPluginCall) {
        print("[AuthenticationPlugin] clearAuthentication called")
        ensureConfigured()
        
        do {
            try TruvideoSdk.clearAuthentication()
            print("[AuthenticationPlugin] clearAuthentication success")
            call.resolve(["clearAuthentication": "Clear success"])
        } catch {
            print("[AuthenticationPlugin] clearAuthentication failed: \(error.localizedDescription)")
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func toSha256String(_ call: CAPPluginCall) {
        print("[AuthenticationPlugin] toSha256String called")
        
        guard let secretKey = call.getString("secretKey"),
              let payload = call.getString("payload") else {
            print("[AuthenticationPlugin] Error: Missing secretKey or payload")
            call.reject("Missing secretKey or payload")
            return
        }
        
        guard let keyData = secretKey.data(using: .utf8),
              let payloadData = payload.data(using: .utf8) else {
            print("[AuthenticationPlugin] Error: Failed to encode secretKey or payload")
            call.reject("Invalid secretKey or payload encoding")
            return
        }
        
        var hmac = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        keyData.withUnsafeBytes { keyBytes in
            payloadData.withUnsafeBytes { payloadBytes in
                CCHmac(CCHmacAlgorithm(kCCHmacAlgSHA256), keyBytes.baseAddress, keyData.count, payloadBytes.baseAddress, payloadData.count, &hmac)
            }
        }
        
        let hash = hmac.map { String(format: "%02x", $0) }.joined()
        print("[AuthenticationPlugin] Generated SHA256 Hash: \(hash)")
        
        call.resolve(["signature": hash])
        print("[AuthenticationPlugin] SHA256 signature sent to JS")
    }
}

File: android/src/main/java/com/truvideo/plugin/core/AuthenticationPlugin.java
package com.truvideo.plugin.core;


import androidx.annotation.NonNull;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.truvideo.sdk.core.TruvideoSdk;
import com.truvideo.sdk.core.interfaces.TruvideoSdkCallback;
import com.truvideo.sdk.model.exceptions.TruvideoSdkException;

import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import kotlin.Unit;

@CapacitorPlugin(name = "Authentication")
public class AuthenticationPlugin extends Plugin {

    private static final String TAG = "AuthenticationPlugin";
    private static final int OTP_HTTP_CONNECT_MS = 15_000;
    private static final int OTP_HTTP_READ_MS = 15_000;

    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    private final Authentication implementation = new Authentication();

    private void runOnMainThread(Runnable runnable) {
        mainHandler.post(runnable);
    }

    private static boolean isBlank(String s) {
        return s == null || s.trim().isEmpty();
    }

    /** Trims trailing `/` characters (matches Kotlin trimEnd('/')). */
    private static String trimTrailingSlashes(String baseUrl) {
        if (isBlank(baseUrl)) {
            return "";
        }
        int end = baseUrl.length();
        while (end > 0 && baseUrl.charAt(end - 1) == '/') {
            end--;
        }
        return baseUrl.substring(0, end);
    }

    /** HMAC-SHA256 hex, UTF-8 key and payload — same semantics as OTP generate RN bridge. */
    private static String hmacSha256Hex(String secret, String payload) throws NoSuchAlgorithmException, InvalidKeyException {
        Mac mac = Mac.getInstance("HmacSHA256");
        SecretKeySpec keySpec = new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
        mac.init(keySpec);
        byte[] digest = mac.doFinal(payload.getBytes(StandardCharsets.UTF_8));
        StringBuilder hex = new StringBuilder(digest.length * 2);
        for (byte b : digest) {
            hex.append(String.format(java.util.Locale.US, "%02x", b & 0xff));
        }
        return hex.toString();
    }

    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");

        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }



    @PluginMethod
    public void isAuthenticated(PluginCall call){
        Boolean isAuth = TruvideoSdk.getInstance().isAuthenticated();
        JSObject ret = new JSObject();
        Log.i("Echo", "isAuthenticated");
        ret.put("isAuthenticated", implementation.echo(isAuth.toString()));
        call.resolve(ret);
    }

    @PluginMethod
    public void version(PluginCall call){
        String isAuth = TruvideoSdk.getInstance().getVersion();
        JSObject ret = new JSObject();
        Log.i("Echo", "version");
        ret.put("version", implementation.echo(isAuth));
        call.resolve(ret);
    }

    @PluginMethod
    public void getApiKey(PluginCall call){
        // getApiKey is not available in the current SDK version.
        String apiKey = "";
        JSObject ret = new JSObject();
        Log.i("Echo", "apikey");
        ret.put("apiKey", implementation.echo(apiKey));
        call.resolve(ret);
    }

    @PluginMethod
    public void environment(PluginCall call){
        String env = TruvideoSdk.getInstance().getEnvironment().name();
        JSObject ret = new JSObject();
        Log.i("Echo", "environment");
        ret.put("environment", implementation.echo(env));
        call.resolve(ret);
    }

    @PluginMethod
    public void isAuthenticationExpired(PluginCall call){
        // isAuthenticationExpired is not available in the current SDK version.
        // We fallback to checking if not authenticated.
        boolean isExpired = !TruvideoSdk.getInstance().isAuthenticated();
        JSObject ret = new JSObject();
        Log.i("Echo", "isAuthenticationExpired");
        ret.put("isAuthenticationExpired", implementation.echo(String.valueOf(isExpired)));
        call.resolve(ret);
    }

    @PluginMethod
    public void generatePayload(PluginCall call){
        String generatePayload = TruvideoSdk.getInstance().generatePayload();
        JSObject ret = new JSObject();
        Log.i("Echo", "generatePayload");
        ret.put("generatePayload", implementation.echo(generatePayload));
        call.resolve(ret);
    }

    @PluginMethod
    public void authenticate(PluginCall call){
        String apiKey = call.getString("apiKey");
        String payload = call.getString("payload");
        String signature = call.getString("signature");
        String externalId = call.getString("externalId");
        if(apiKey == null || payload == null || signature == null || externalId == null){
            return;
        }
        Log.i("Echo", "authenticate call ");
        TruvideoSdk.getInstance().authenticate(apiKey, payload, signature,externalId,
                new TruvideoSdkCallback<Unit>(){
                    @Override
                    public void onComplete(Unit unit) {
                        JSObject ret = new JSObject();
                        Log.i("Echo", "authenticate");
                        ret.put("authenticate", implementation.echo("Authentication success"));
                        call.resolve(ret);
                    }
                    @Override
                    public void onError(@NonNull TruvideoSdkException e) {
                        // handle error
                        Log.i("Echo", "authenticate fail");
                        call.reject(e.toString());
                    }
                });

    }

    @PluginMethod
    public void initAuthentication(PluginCall call){
        Log.i("Echo", "initAuthentication call");
        TruvideoSdk.getInstance().waitAuthReady(new TruvideoSdkCallback<Unit>() {
            @Override
            public void onComplete(Unit unit) {
                // Authentication ready
                JSObject ret = new JSObject();
                Log.i("Echo", "initAuthentication");
                ret.put("initAuthentication", implementation.echo("Init success"));
                call.resolve(ret);
            }

            @Override
            public void onError(@NonNull TruvideoSdkException e) {
                // handle error
                Log.i("Echo", "initAuthentication fail");
                call.reject(e.toString());

            }
        });

    }

    @PluginMethod
    public void clearAuthentication(PluginCall call){
        TruvideoSdk.getInstance().clearAuthentication(new TruvideoSdkCallback<Unit>() {
            @Override
            public void onComplete(Unit unit) {
                JSObject ret = new JSObject();
                ret.put("clearAuthentication", implementation.echo("Clear success"));
                call.resolve(ret);
            }

            @Override
            public void onError(@NonNull TruvideoSdkException e) {
                call.reject(e.toString());
            }
        });
    }

    @PluginMethod
    public void toSha256String(PluginCall call){
        try {
            // getting instance of Message Authentication Code
            String secretKey = call.getString("secretKey");
            String payload = call.getString("payload");
            if(secretKey == null || payload == null){
                return;
            }
            Mac hmacSha256 = Mac.getInstance("HmacSHA256");
            //secretKey
            SecretKeySpec secret = new SecretKeySpec(secretKey.getBytes(), "HmacSHA256");
            hmacSha256.init(secret);
            byte[] macData = hmacSha256.doFinal(payload.getBytes());

            // Convert byte array to hex string
            StringBuilder hexString = new StringBuilder();
            for (byte b : macData) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) {
                    hexString.append('0');
                }
                hexString.append(hex);
            }
            JSObject ret = new JSObject();
            Log.i("Echo", "toSha256String");
            ret.put("signature", implementation.echo(hexString.toString()));
            call.resolve(ret);
        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
            e.printStackTrace();
            call.reject(e.toString());
        }

    }

    /**
     * Pure HTTP OTP generation (same contract as React Native {@code generateOtp}).
     */
    @PluginMethod
    public void generateOtp(PluginCall call) {
        String baseUrl = call.getString("baseUrl");
        String apiKey = call.getString("apiKey");
        String secret = call.getString("secret");
        String externalId = call.getString("externalId");

        if (isBlank(apiKey)) {
            call.reject("apiKey cannot be empty", "OTP_GENERATE_ERROR", (Exception) null);
            return;
        }
        if (isBlank(secret)) {
            call.reject("secret cannot be empty", "OTP_GENERATE_ERROR", (Exception) null);
            return;
        }
        if (isBlank(externalId)) {
            call.reject("externalId cannot be empty", "OTP_GENERATE_ERROR", (Exception) null);
            return;
        }
        if (isBlank(baseUrl)) {
            call.reject("baseUrl cannot be empty", "OTP_GENERATE_ERROR", (Exception) null);
            return;
        }

        Log.i(TAG, "generateOtp request started");

        new Thread(() -> {
            HttpURLConnection connection = null;
            try {
                String cleanBase = trimTrailingSlashes(baseUrl);
                String endpoint = cleanBase + "/api/v1/auth/otp/generate";
                String body = new JSONObject().put("externalId", externalId).toString();
                String signature = hmacSha256Hex(secret, body);

                URL url = new URL(endpoint);
                connection = (HttpURLConnection) url.openConnection();
                connection.setRequestMethod("POST");
                connection.setConnectTimeout(OTP_HTTP_CONNECT_MS);
                connection.setReadTimeout(OTP_HTTP_READ_MS);
                connection.setDoOutput(true);
                connection.setRequestProperty("Content-Type", "application/json");
                connection.setRequestProperty("x-authentication-api-key", apiKey);
                connection.setRequestProperty("x-authentication-signature", signature);

                try (OutputStream os = connection.getOutputStream()) {
                    os.write(body.getBytes(StandardCharsets.UTF_8));
                }

                int status = connection.getResponseCode();
                InputStream stream =
                        status >= 200 && status <= 299
                                ? connection.getInputStream()
                                : connection.getErrorStream();
                String responseText = readStreamAsString(stream);

                if (status < 200 || status > 299) {
                    String apiMsg = parseJsonErrorMessage(responseText);
                    String err =
                            !isBlank(apiMsg)
                                    ? ("OTP generate failed (" + status + "): " + apiMsg)
                                    : ("OTP generate failed with status " + status);
                    throw new IllegalStateException(err);
                }

                String otp;
                try {
                    otp = new JSONObject(responseText).optString("otp", "");
                } catch (Exception parseEx) {
                    otp = "";
                }
                if (isBlank(otp)) {
                    throw new IllegalStateException("OTP not found in response");
                }

                JSObject ret = new JSObject();
                ret.put("generateOtp", otp);
                runOnMainThread(() -> call.resolve(ret));
                Log.i(TAG, "generateOtp success");
            } catch (Exception e) {
                Log.e(TAG, "generateOtp failed", e);
                runOnMainThread(() -> call.reject(e.toString(), "OTP_GENERATE_ERROR", e));
            } finally {
                if (connection != null) {
                    connection.disconnect();
                }
            }
        }, "generateOtp").start();
    }

    /**
     * OTP login via native SDK (matches React Native {@code authenticateWithOtp} flow:
     * {@code authenticate(otp)} then {@code waitAuthReady()}).
     */
    @PluginMethod
    public void authenticateWithOtp(PluginCall call) {
        String otp = call.getString("otp");
        if (isBlank(otp)) {
            call.reject("OTP cannot be empty", "OTP_AUTH_ERROR", (Exception) null);
            return;
        }
        final String trimmedOtp = otp.trim();
        Log.i(TAG, "authenticateWithOtp called");

        TruvideoSdk.getInstance()
                .authenticate(
                        trimmedOtp,
                        new TruvideoSdkCallback<Unit>() {
                            @Override
                            public void onComplete(@NonNull Unit unit) {
                                TruvideoSdk.getInstance()
                                        .waitAuthReady(
                                                new TruvideoSdkCallback<Unit>() {
                                                    @Override
                                                    public void onComplete(@NonNull Unit unit) {
                                                        JSObject ret = new JSObject();
                                                        ret.put(
                                                                "authenticateWithOtp",
                                                                "OTP Authentication Successful");
                                                        runOnMainThread(() -> call.resolve(ret));
                                                        Log.i(TAG, "authenticateWithOtp success");
                                                    }

                                                    @Override
                                                    public void onError(
                                                            @NonNull TruvideoSdkException e) {
                                                        Log.e(TAG, "waitAuthReady error", e);
                                                        runOnMainThread(
                                                                () ->
                                                                        call.reject(
                                                                                e.toString(),
                                                                                "OTP_AUTH_ERROR",
                                                                                e));
                                                    }
                                                });
                            }

                            @Override
                            public void onError(@NonNull TruvideoSdkException e) {
                                Log.e(TAG, "authenticate(otp) error", e);
                                runOnMainThread(
                                        () -> call.reject(e.toString(), "OTP_AUTH_ERROR", e));
                            }
                        });
    }

    private static String readStreamAsString(InputStream stream) throws IOException {
        if (stream == null) {
            return "";
        }
        BufferedReader reader =
                new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            sb.append(line);
        }
        reader.close();
        return sb.toString();
    }

    /**
     * Best-effort extraction of {@code message} or {@code detail} from error JSON bodies.
     */
    private static String parseJsonErrorMessage(String responseText) {
        if (isBlank(responseText)) {
            return "";
        }
        try {
            JSONObject jo = new JSONObject(responseText);
            String message = jo.optString("message", "");
            if (!isBlank(message)) {
                return message.trim();
            }
            String detail = jo.optString("detail", "");
            return isBlank(detail) ? "" : detail.trim();
        } catch (Exception e) {
            return "";
        }
    }

}
